src/config_inference.ts

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021 Deno Land Inc. All rights reserved. MIT license.
+
+import { basename, magenta } from "../deps.ts";
+import { API, APIError } from "./utils/api.ts";
+import TokenProvisioner from "./utils/access_token.ts";
+import { wait } from "./utils/spinner.ts";
+import { error } from "./error.ts";
+
+const NONAMES = ["src", "lib", "code", "dist", "build", "shared", "public"];
+
+/** Arguments inferred from context */
+interface InferredArgs {
+  project?: string;
+  entrypoint?: string;
+}
+
+/**
+ * Infer name of the project.
+ *
+ * The name of the project is inferred from either of the following options, in order:
+ * - If the project is in a git repo, infer `<org-name>-<repo-name>`
+ * - Otherwise, use the directory name from where DeployCTL is being executed,
+ *   unless the name is useless like "src" or "dist".
+ */
+async function inferProject(api: API) {
+  let projectName = await inferProjectFromOriginUrl() ||
+    inferProjectFromCWD();
```

Comment:
Does `??` also work here instead of `||`? If it does then I think `??` is a little bit better because `??` only works for null-ish values whereas `||` involves implicit type conversion (which makes code reading a bit harder, for me at least)

Reply:
We want to filter out also empty strings as well, hence the `||` usage. We could be explicit about it, but feels a bit unnecessary? verbosity also makes reading code harder.

Reply:
OK, in case that we want to filter out empty strings here, using `||` makes sense to me.